### ISSUE TITLE
Segmentation fault error

### ISSUE THREAD

**Issue by gordonmoon** [tmsimont/wombat#1]
At the end of executions for both CPU and GPU implementations, segmentation fault errors are occurred.
I used text8 dataset and intel/17.0.7 and cuda/9.2.88 compliers.

**Comment by tmsimont** [comment]
I refactored a lot of the code after getting my initial results... unfortunately I've kind of neglected this since.

I wonder if maybe resetting to the initial commit wouldn't work for you.

If you do:

```
git reset --hard aa1d0fbdbf7d342143c25484f88e2063aa025460
```

in your repo... and then compile. does that work?

In the mean time I'll see if I can reproduce the problem.

**Comment by tmsimont** [comment]
Actually -- before you do that...

How are you running the program? Are you using one of the scripts? What params are you giving it?

**Comment by tmsimont** [comment]
for what it's worth I don't see any issues using g++ on my local arch box but I sadly don't have access to the Intel and CUDA machines I used to develop this. I'm looking at AWS instances that might do the job.. I've been meaning to clean up this project

**Comment by tmsimont** [comment]
It also works on my local nvidia card:

```
$ ./cuda.numCPUT-batchSize-batchesPerT.sh 4 125 100
num threads: 4
num physical cores: 2
tcbs per thread: 1
items in tcb: 125
hs: 1
number of iterations: 10
hidden size: 100
number of negative samples: 5
window size: 8
starting learning rate: 0.02500
starting training using file: ../data/text8

Vocab size: 71291
Words in train file: 16718843
Alpha: 0.000032  Progress: 99.88%  Words/sec: 362.59k
Final: Alpha: 0.000030  Progress: 99.88%  Words/sec: 361.99k
```

That's not saying much. I do wonder if maybe the input params you're giving aren't the issue.

I remember running into some weird issues when using various combinations of "batch size" and "batches per thread"

I will look into it a bit more and see if can't clean up the docs around those params.

**Comment by gordonmoon** [comment]
Thanks for your effort on running your code. However when I tried with the same parameters as shown above, it still returns an error. The below is the result I have gotten.

-bash-4.2$ ./cuda.numCPUT-batchSize-batchesPerT.sh 4 125 100
num threads: 4
num physical cores: 2
tcbs per thread: 1
items in tcb: 125
hs: 1
number of iterations: 10
hidden size: 100
number of negative samples: 5
window size: 8
starting learning rate: 0.02500
starting training using file: ../data/text8

Vocab size: 71291
Words in train file: 16718843
Alpha: 0.000031  Progress: 99.88%  Words/sec: 2103.93k
Final: Alpha: 0.000029  Progress: 99.88%  Words/sec: 2103.38k
./cuda.numCPUT-batchSize-batchesPerT.sh: line 16: 159900 Segmentation fault      ../wombat -train ../data/text8 -output ../output/vectors-cuda-$1$2$3.txt -binary 0 -cbow 0 -size 100 -window 8 -sample 1e-4 -negative 5 -hs 1 -iter 10 -num-threads $1 -tcbs-per-thread 1 -items-in-tcb 125 -batch-size $2 -batches-per-thread $3 -senbs 1 -debug 2 -time-events 0

**Comment by gordonmoon** [comment]
Also, here is what I have gotten by executing CPU implementation.

-bash-4.2$ ./cpu.sh 4
num threads: 4
num physical cores: 4
tcbs per thread: 1
items in tcb: 125
hs: 1
number of iterations: 10
hidden size: 100
number of negative samples: 5
window size: 8
starting learning rate: 0.02500
starting training using file: ../data/text8

Vocab size: 71291
Words in train file: 16718843
Alpha: 0.000040  Progress: 99.85%  Words/sec: 2049.89k
Final: Alpha: 0.000039  Progress: 99.85%  Words/sec: 2049.52k
./cpu.sh: line 13: 84889 Segmentation fault      ../wombat -train ../data/text8 -output ../output/vectors-cpu-$1.txt -binary 0 -cbow 0 -size 100 -window 8 -sample 1e-4 -negative 5 -hs 1 -iter 10 -num-threads $1 -debug 2 -num-phys $1

**Comment by gordonmoon** [comment]
Could you let me know your machine configurations for both CPUs and GPUs?

**Comment by tmsimont** [comment]
```
$ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2018 NVIDIA Corporation
Built on Tue_Jun_12_23:07:04_CDT_2018
Cuda compilation tools, release 9.2, V9.2.148

$ g++ -v
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/8.1.1/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /build/gcc/src/gcc/configure --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --enable-languages=c,c++,ada,fortran,go,lto,objc,obj-c++ --enable-shared --enable-threads=posix --enable-libmpx --with-system-zlib --with-isl --enable-__cxa_atexit --disable-libunwind-exceptions --enable-clocale=gnu --disable-libstdcxx-pch --disable-libssp --enable-gnu-unique-object --enable-linker-build-id --enable-lto --enable-plugin --enable-install-libiberty --with-linker-hash-style=gnu --enable-gnu-indirect-function --enable-multilib --disable-werror --enable-checking=release --enable-default-pie --enable-default-ssp
Thread model: posix
gcc version 8.1.1 20180531 (GCC) 
```

can you edit the `cpu.sh` script and make the first line into:

```
gdb --args ../wombat \
```

That should fire up gdb...

you then just type `r` -> return

that will start the program in gdb, which might reveal more information about where this seg fault is coming from.

I've been a little busy at work lately and haven't had a chance to try this on a machine other than my laptop.

**Comment by tmsimont** [comment]
I kind of suspect the problem is coming from this code: 

https://github.com/tmsimont/wombat/blob/4da18f3a196b6ccf7f1b20483bd1fc39678b633e/src/w2v-functions.cpp#L400-L416

I'm just not sure how and why only for you and not me :p

**Comment by tmsimont** [comment]
ah one more question -- are you using `make intel` or just `make`? I wonder if the intel compiler isn't behaving a little differently.

**Comment by gordonmoon** [comment]
I used make intel and make cuda to compile your code.
After I comment out saveModel(), the function calls in both main.cpp and main-coda.cpp, there are no more segmentation fault errors. Since I don't need to save trained parameters, this issue is not problematic for me.
Thank you so much for your time and effort on resolving the problem.

**Comment by tmsimont** [comment]
No problem! I bet it's caused by some kind of file permissions issue. It's probably unable to open the output file for writing and failing. 

Good luck with your work!

**Comment by mateusffreitas** [comment]
The problem is  -output ../**output**/vectors-cpu-$1.txt
The folder does not exist.
Remember to check file pointers for NULL.

**Comment by tmsimont** [comment]
ah yes makes sense. i need to put a .gitkeep in that folder so it's there when you clone the repo.

i've started cleaning up the code in this project but it's a big task :)
